File: algorithms/[1734]解码异或后的排列/solution.ts
// topics = ["位运算"]

/**
 * Bit Manipulation
 * time O(n), space O(1), n 为原始数组 perm 的长度
 */
function decode(encoded: number[]): number[] {
  const n = encoded.length + 1;

  // 原始数组所有元素的异或结果
  let total = 0;
  for (let i = 1; i <= n; i += 1) {
    total ^= i;
  }

  // 原始数组除第一个元素外，其它所有元素的异或结果(n 为奇数！)
  let others = 0;
  for (let i = 1; i < n; i += 2) {
    others ^= encoded[i];
  }

  const res = [total ^ others];
  for (const v of encoded) {
    res.push(res[res.length - 1] ^ v);
  }

  return res;
}

File: algorithms/[1734]解码异或后的排列/solution.py
# topics = ["位运算"]


from typing import List
from operator import xor
from functools import reduce


class Solution:
    def decode(self, encoded: List[int]) -> List[int]:
        """
        Bit Manipulation
        time O(n), space O(1), n 为原始数组 perm 的长度
        """
        n = len(encoded) + 1
        total = reduce(xor, range(1, n + 1))
        odd = 0
        for i in range(1, n - 1, 2):
            odd ^= encoded[i]

        perm = [total ^ odd]
        for i in range(n - 1):
            perm.append(perm[-1] ^ encoded[i])

        return perm
